Edit Security Guide's Session Guidelines [skip ci]

matildasmeds opened the pull request:
* Fix `rails secret` -> `rake secret`
* For increased focus and ease of reading
** Remove mentions related to earlier Rails versions
** Condense the section where it seems viable

### Summary

Reading Security Guide, I noticed that `rails secret` either does not exist, or I don't know how it works, but `rake secret` does exist and works as described in the guide. So that is changed.

Additionally, I have made an attempt to make "Session Guidelines" easy and to-the-point read to someone who wants to understand web security concepts and get going with Rails 5. 

Removing mentions related to earlier Rails versions seems like a sensible clean up, because someone using earlier Rails versions would look at the corresponding Rails Guide. Then ideally, Rails Guides would cover the version they are meant for, not more, and not less. Does this idea match with the vision for Rails Guides?

I have condensed the section somewhat, to make it easier to digest. Finding the balance between giving enough information, but still keeping the text pruned enough for a clear read is a subjective matter. :)

### Other Information

This is my first PR, I am not a security expert, and look forward to any discussion regarding the commit, and Rails Guides in general.

The diff reads best in Split mode.

matildasmeds committed:
xEdit Security Guide's Session Guidelines section [skip ci]
* For topicality, focus, and ease of reading
** Remove mentions related to earlier Rails versions
** Remove and condense perhaps tangential parts
** Link to Action Controller Guide, when appropriate

rafaelfranca commented:
`rails secret` does exist since Rails 5. Could you change it back?

matildasmeds commented:
Fixed. Sorry about that.

rafaelfranca commented:
> E: How does this work, should I update the original comment to reflect latest content of the diff?

You don't need to.

The patch looks good to me. @fxn @matthewd can you review it too?

matthewd commented on guides/source/security.md:
```diff
@@ -80,31 +80,35 @@ The main objective of most attackers is to make money. The underground prices fo
-NOTE: _Rails provides several storage mechanisms for the session hashes. The most important is `ActionDispatch::Session::CookieStore`._
+NOTE: _Rails provides several storage mechanisms for the session hashes. The most important one is `ActionDispatch::Session::CookieStore`._
+
+Rails `ActionDispatch::Session::CookieStore` saves the session hash in a cookie on the client-side. The server retrieves the session hash from the cookie, which effectively acts as the session ID. When you use cookies as the session storage, it is good to understand how they work.
```
It really doesn't act *as* a session ID -- it just makes one unnecessary.

Lose the "Rails" at the start of this line; it would need to be possessive, and that seems needlessly awkward when the class name can stand alone.

matthewd commented on guides/source/security.md:
```diff
@@ -80,31 +80,35 @@ The main objective of most attackers is to make money. The underground prices fo
+
+Rails `ActionDispatch::Session::CookieStore` saves the session hash in a cookie on the client-side. The server retrieves the session hash from the cookie, which effectively acts as the session ID. When you use cookies as the session storage, it is good to understand how they work.
+
+* Cookies imply a strict size limit of 4kB. Use the cookie only for data which is relevant for the session.
```
imply -> have ?

matthewd commented on guides/source/security.md:
```diff
@@ -80,31 +80,35 @@ The main objective of most attackers is to make money. The underground prices fo
+
+* Cookies imply a strict size limit of 4kB. Use the cookie only for data which is relevant for the session.
+
+* The client has full read and write access to the cookie. Client can copy cookies from computer to another. _In general, do not store anything sensitive or security related in the session._
```
Not true; it's encrypted and signed

matthewd commented on guides/source/security.md:
```diff
@@ -80,31 +80,35 @@ The main objective of most attackers is to make money. The underground prices fo
+* When you have `secret_token` set, Rails is able to sign the cookie. The client can read the cookie, but cannot alter contents without Rails detecting it.
 
-That means the security of this storage depends on this secret (and on the digest algorithm, which defaults to SHA1, for compatibility). So _don't use a trivial secret, i.e. a word from a dictionary, or one which is shorter than 30 characters, use `rails secret` instead_.
+* When you have `secret_key_base` set, Rails is able to encrypt the cookie. The client cannot read or alter the cookie, without breaking the encryption.
```
This and the above can be rephrased a bit: Rails refuses to start if it doesn't have a `secret_key_base`, so no need for conditionals.

matthewd commented on guides/source/security.md:
```diff
@@ -80,31 +80,35 @@ The main objective of most attackers is to make money. The underground prices fo
+#### Security measures to protect your cookies
 
-* Cookies imply a strict size limit of 4kB. This is fine as you should not store large amounts of data in a session anyway, as described before. _Storing the current user's database id in a session is usually ok_.
+Setup secrets in config/secret.yml for all your environments.
```
secrets.yml, plural

matthewd commented on guides/source/security.md:
```diff
@@ -80,31 +80,35 @@ The main objective of most attackers is to make money. The underground prices fo
+Setup secrets in config/secret.yml for all your environments.
 
-* The client can see everything you store in a session, because it is stored in clear-text (actually Base64-encoded, so not encrypted). So, of course, _you don't want to store any secrets here_. To prevent session hash tampering, a digest is calculated from the session with a server-side secret (`secrets.secret_token`) and inserted into the end of the cookie.
+When choosing secrets, make sure they are long and random enough, and consequently, harder to break. _Do not use a trivial secrets, i.e. a dictionary word, or secrets shorter than 30 characters_.
```
Maybe simpler to collapse this and below down to 'secrets must be long and random. use `rails secret`.'

matthewd commented on guides/source/security.md:
```diff
@@ -115,9 +119,9 @@ That means the security of this storage depends on this secret (and on the diges
+Use `rails secret` command to get strong random secrets you can use. Use different secrets for different environments.
 
-If you have received an application where the secret was exposed (e.g. an application whose source was shared), strongly consider changing the secret.
+If an application's secret may have been exposed, strongly consider changing the secret. Changing the secret will expire currently active sessions.
```
I think it might be worth keeping the "source was shared" example here: beyond the direct utility of pointing out a common way for the secret to be compromised, it provides some context on the *sorts* of things that will expose the secret -- and by implication, things that probably won't.

matthewd reviewed (commented):
This looks great, thanks!

matildasmeds commented:
Thanks for the comments. I have attempted to address all of them. Here's the updated, and perhaps a slightly more condensed version. 

Action Controller Overview Guide has a good explanation on how to use sessions in general, and I felt Security Guide should rather refer to that, rather than have separate "Session guideline" section. 

Some overlap is perhaps even desirable, but keeping in mind the reader would also read the Action Controller Overview (especially if we remind them about it in the Security Guide), some overlap could be removed.

fxn commented:
This stuff was already present in the guide, but since the patch is touching this let me add some remarks.

The intro paragraph talks about session identifiers, but with cookie-based sessions there are no session IDs, and this is the default in Rails. So the section opens with something that is not true for the vast majority of Rails applications, which makes no sense. Session ID and database storage should have an exceptional rather than prominent treatment.

Also, the guide claims that sessions make HTTP stateful, which is generally not true. A cookie-based session is tunneling request parameters via a HTTP header. The request is self-contained, there is no session state in the server.

matildasmeds commented:
Thanks for the comments @fxn! Do you think if these claims are corrected the PR could be accepted?

matildasmeds commented:
Thanks for comments! Here would be the updated version. 

Changes explained below:
* Removed the "stateful HTTP" comment altogether.
* Removed the session method example, instead linked to the Action Controller Overview Guide, because there are examples already available there.

I kept an edited reference to session ID:s for two reasons.  
1. It looks like Cookie Store and Cache Store use session_id's, even in Rails 5. Looking at Rails source, both `ActionDispatch::Session::CookieStore` and `ActionDispatch::Session::CacheStore` seem to use session id (`sid`) to identify the session. I have a toy Rails 5.0.1 app, using default session storage, and session object is initialized with session_id key, whose value stays the same between requests. I assume the session_id is stored on the cookie, however I did not decrypt the cookie to verify this. If I am missing something here, please let me know.
2. Mentioning session id seems like good background for the "Session hijacking" section, and the introductory section might be a good place for that.

Looking forward to any comments :)

matildasmeds commented:
@rafaelfranca Please let me know if there something more I can do, to move this PR forward.

oniofchaos commented:
@matildasmeds noticing this never got merged and this now has a merge conflict unfortunately. Would you be able to clean it up so we can try and get it merged?

matildasmeds commented:
@oniofchaos Nice to hear from you. If there is a chance for this PR getting merged, I can definitely take a look, and try to resolve the merge conflict in a nice way.

rafaelfranca commented:
Definitly wants to get this merged. There were a few changes since the PR was opened though. I think the main one is that config/credentials is the preferred way to set the secret key. Do you want to update it?

matildasmeds committed:
Merge branch 'master' into guides_session_guidelines

matildasmeds commented:
Hi @rafaelfranca, @oniofchaos, here's my try at resolving the merge conflicts. 
With regard to config/secrets.yml I removed the part about config/secrets, and linked from that section to a later section in the same guide, which describes config/credentials. If there are further suggestions for this PR, please let me know.

matildasmeds commented:
Hi @rafaelfranca & @oniofchaos! Here's my attempt at resolving the merge conflicts. With regard to config/secrets, I removed that part, and linked from there to other part of the same guide, where config/credentials is described. While the credentials part could be a bit clearer, I would prefer to look at that in another PR. If there are any further comments for this PR, please let me know. :)

matildasmeds commented:
This version is made with GitHub resolve conflicts editor, and there's a [long section on encrypted session storage](https://github.com/rails/rails/pull/28686/commits/c47ab8702762047b48c2d2424a98425003a5d59f#diff-2b3e99b648bf1669cc6d10cd6cb747e1R96) which is not written by me, although in the diff it looks like it is. Would it have been better to make the edits locally?

oniofchaos commented on guides/source/security.md:
```diff
@@ -32,21 +32,17 @@ In order to develop secure web applications you have to keep up to date on all l
 ### What are Sessions?
 
-NOTE: _HTTP is a stateless protocol. Sessions make it stateful._
+INFO: Sessions enable the application to maintain user-specific state, while users interact with the application. Once the user has been authenticated, an active session enables the user keep using the application, without authenticating on each request.
```
`an active session enables the user` _to_ `keep using the application` and also remove the comma before `without` (doesn't seem necessary)?

Thoughts?

oniofchaos commented on guides/source/security.md:
```diff
@@ -76,34 +72,28 @@ Hence, the cookie serves as temporary authentication for the web application. An
-### Encrypted Session Storage
+* Cookies are stored on the client-side. The client may preserve cookie contents even for expired cookies. The client may copy cookies to other machines. Avoid storing sensitive data in cookies.
+
+* Cookies are temporary by nature. The server can set expiration time for the cookie, but the client may delete the cookie and its contents before that. Persist the data you want to persist on the server side.
```
> Persist the data you want to persist on the server side.

Double usage of `persist` here feels iffy, maybe something like `Persist stateful data on the server side.` or `Persist data on the server side that needs to remain between visits.`

Thoughts?

dillonwelch reviewed (commented):
Just some potential wording changes to make. Diff seems small enough still to keep within one PR IMO, but I'm not a maintainer 😇

matildasmeds commented:
Thanks @oniofchaos for the comments, I really appreciate it and would be happy to revisit these wordings. :mag: :sweat_smile:   

Personally I'm not super comfortable with this diff, because I feel like I am signing off content, which I haven't written, nor reviewed. I would prefer to open a new PR in any case. 

So how about I open one new PR, include the contents from this one + credentials thing, and close this one? We can still keep everything in one PR if it is easier to review that way. I would summon up my inner Hemingway, and try to get it done over the weekend. Wishing you a good day!

oniofchaos commented:
You're the boss. Feel free to tag me in new PRs and I'll edit your inner Hemingway.
